### DISCUSSION TITLE
Quicklook banner button tapped event not working for URL redirect

### DISCUSSION THREAD

**Discussion by karthickRamadoss** [google/model-viewer#2755] [Q&A]
Quicklook banner button tapped event not working for URL redirect, Please find code snippet 
**### html  model viewer code** 
 ios- src="customIos.usdz#applePayButtonType=buy&checkoutTitle=Biplane%20Toy&checkoutSubtitle=Rustic%20finish%20with%20rotating%20propeller&price=$15"

**### Java script callback for banner action** 
const linkElementNew = document.getElementById("customID");
linkElementNew.addEventListener("message", function (event) {   
    if (event.data == "_apple_ar_quicklook_button_tapped") {
        window.location.assign('https://www.youtube.com/')
    }
}, false);

It always returns back to previous page .Any  suggestions for this issue.

**Comment by elalish** [comment]
I have no idea, but it sounds more like a Quick Look / Safari issue to me.

**Reply by elalish** [reply]
@grorg Any thoughts?

**Reply by btutal** [reply]
@elalish actually I am having the same problem. I have tried to use `quick-look-button-tapped` event but couldn't make it work.  You can check live demo [here](https://glitch.com/edit/#!/broad-nimble-fan) When you check it live on an iOS device, you will see a banner but when you press on the banner, you are not redirected to the Model Viewer Documentation page.  

Did anything change regarding this event with iOS 16 or model-viewer 2.0 ?

**Reply by elalish** [reply]
Not that I'm aware of; where is the live demo link?

**Reply by btutal** [reply]
[Here it is](https://glitch.com/edit/#!/broad-nimble-fan)

**Reply by elalish** [reply]
Try pinning your MV version to one that worked before (maybe 1.12?) to test if it's our regression. If not, please file a bug on WebKit.

**Reply by btutal** [reply]
Small update, I have tested from 1.12.1 to 2.0.2 and all seem to work fine on iOS 15.5 and 15.7 and they don't work on iOS 16.0, 16.0.2, and 16.1. I will also try to make some tests with standard implementation from [Apple](https://developer.apple.com/documentation/arkit/adding_an_apple_pay_button_or_a_custom_action_in_ar_quick_look) but it is taking some time since I don't have an access to devices with different iOS versions and don't really know if there is any way to test these with the simulator.

**Reply by btutal** [reply]
@elalish I have updated the demo with the default Apple example and it seems to work fine on all iOS versions which I mentioned above but the MV one doesn't work on iOS 16. Any proposal on how to investigate further? 

[Here is code](https://glitch.com/edit/#!/broad-nimble-fan)
[Live Example](https://broad-nimble-fan.glitch.me)

**Reply by elalish** [reply]
Thanks for the investigation! @grorg, any breaking changes we should know about in iOS16? @btutal, please file an issue on WebKit and link it here; we're more likely to get help from Apple there than here.

**Reply by milesgreen** [reply]
Poking around, here's what I can determine...

Looks like in iOS16, Safari is not propagating the `_apple_ar_quicklook_button_tapped` message event to the anchor element when it's not attached to the DOM. 

See demo here:
https://ios16-arql-banner-issue.glitch.me/

First example; anchor element is defined on the page. Interaction with ARQL banner is captured through message event listener. This works in iOS16 fine.

Second example; the 'Launch Default' button triggers the `click()` method of the default anchor. ARQL Opens. Interaction with banner is captured. Works fine in iOS16.

Third example; **Similar to ModelViewer approach when launching ARQL**. The 'Launch Detached' button uses JavaScript to create a new anchor element, assigns necessary attributes, and triggers `click()` method, _without attaching element to the DOM_. In iOS16, ARQL opens, banner is displayed, but clicks on banner do not propagate the message event to the detached element. This used to (and still does) work fine in iOS15.

Fourth example; The 'Launch Attached' button uses JavaScript to create a new anchor element, but also _attaches it to the DOM_ by appending to `body`. ARQL opens, the banner displays, and banner clicks do indeed get propagated to the anchor's event listener and can respond appropriately. Works in iOS16.


A potential MV fix might be to attach the ARQL anchor to the DOM but hidden. Not sure if there's wider concerns with that?

But the question is; was this an intentional change in behaviour in iOS16 or a bug?

**Reply by elalish** [reply]
I'd guess bug, but as usual it'll be up to us to work around it. I think your solution sounds just fine; want to make a PR since you have the test cases? :pray:

**Reply by milesgreen** [reply]
Sure thing. I'll take a look. 👍

**Reply by milesgreen** [reply]
PR here: https://github.com/google/model-viewer/pull/3850

@btutal @karthickRamadoss - it would be great to know if this demo with the fix now works for you with your variety of test devices:
https://modelviewer-ios16-arql-banner-fix.glitch.me/

**Reply by milesgreen** [reply]
Issue has been filed on WebKit:
https://bugs.webkit.org/show_bug.cgi?id=245959

**Reply by btutal** [reply]
Thanks @milesgreen for further investigation and workaround! :) It works for me on iOS 16 devices as well. On the other hand, @karthickRamadoss might have experienced different problems when this issue is created. I have realized the issue is more than a year old and the code snippet is not using the MV event. OP is not checking here anymore probably :)

**Reply by milesgreen** [reply]
Heh. I didn't notice the original issue was logged a year back. At least we fixed the most recent manifestation. 👍

**Reply by btutal** [reply]
@elalish is there any plan for 2.0.3 or some other release soon or we shouldn't expect any release any time soon and just build our own ?

**Reply by elalish** [reply]
I'm not sure this quite reaches the bar for a point release, especially since it wasn't our regression. If building your own isn't too much onus, I'd appreciate it.

**Comment by milesgreen** [comment]
For future reference, this was addressed with: https://github.com/google/model-viewer/pull/3850
